### PR TITLE
[Linux] Fix ComboBox popup alignment for large scrolling option sets

### PR DESCRIPTION
**Pull request by apman**
Ports the drop-down alignment fix from #652 to the Linux theme.

## Why

`src/Devolutions.AvaloniaTheme.Linux/Controls/ComboBox.axaml` used the same arithmetic `VerticalOffset` estimate as the macOS theme did, and shared its limitations. That estimate only holds while the whole list fits in the popup: once it scrolls, the ComboBox scrolls the selected row into view and the row is no longer at `SelectedIndex * ItemHeight`, so long option sets opened visibly misaligned.

Confirmed rather than assumed — a headless probe on the unmodified Linux theme (1000 items, index 500) perturbed the open popup by 100px and it was never corrected.

## What changed

The template:

- **`ComboBoxPopupAlignmentBehavior.Enable`** — measures the popup's real layout and corrects it, instead of predicting. The behavior is theme-agnostic (it only needs a popup named `PART_Popup`), so every fix made for macOS applies here too: measure-don't-predict, re-arming on `SelectionChanged`, and the scale-agnostic coordinate conversion.
- **`PlacementConstraintAdjustment="SlideX, SlideY, ResizeX, ResizeY`** — suppresses `FlipY`. The default (`All`) includes it, which suits menus but not a drop-down deliberately offset upwards: near the bottom of the screen the compositor flips it as well, leaving it floating a popup-height above the control.

`EditableComboBox` is deliberately untouched — its drop-down is supposed to stay below the control.

### Timing fixes found during on-device testing

The behavior itself needed three corrections that only surfaced against a real compositor. On X11 (which is also what a Wayland session uses, via XWayland — Avalonia has no Wayland backend) the popup's move lands asynchronously via `ConfigureNotify`, well after the dispatcher passes that drive the correction loop. macOS did not expose these because its moves land synchronously.

- **A correction could be applied twice from one measurement.** The pass immediately after a correction usually re-reads the pre-move position. That stale reading only counted towards the "popup is clamped" streak after repeating three times; below that it fell straight through and applied the full correction again. The popup then moved by twice what was needed, so the misalignment returned with the same magnitude and the opposite sign, oscillating until the pass budget ran out. Any unchanged reading is now treated as in-flight.
- **An in-flight move could be mistaken for a clamped popup.** Counting dispatcher passes is not sufficient, as they re-run far faster than the `ConfigureNotify` round-trip; the streak now also has to span real elapsed time.
- **The loop could stop on a not-yet-settled reading.** It unsubscribes as soon as one reading is within tolerance — deliberately, since staying subscribed makes the drop-down fight the user's own scrolling — which left a small permanent residual. A single delayed settle check now re-arms only if the popup actually ended up wrong.

## Tests

Five new tests, and it is worth saying what is *not* there. I first wrote the obvious ones — open the drop-down at various indices, assert the row is aligned. All passed, and all were worthless: they still passed with the behavior fully disabled, and even with the arithmetic estimate deliberately sabotaged, because under the headless overlay popup host the row lands aligned at every index on its own. I swept indices 3/50/100/150/198 looking for a case that would bite; none did, so they were deleted (the same conclusion reached twice in #652).

What remains works by perturbing an *open* drop-down and asserting the loop closes the gap. Each was mutation-tested:

| Test | Mutation that makes it fail |
| --- | --- |
| `Drop_down_is_not_allowed_to_flip_to_the_other_side` | remove `PlacementConstraintAdjustment` |
| `Offset_moved_while_open_is_corrected_without_reopening` | remove the `Enable` setter, or neuter `RunPass` |
| `Stepping_between_clamped_selections_still_re_aligns` | remove the `Enable` setter, or neuter `RunPass` |
| `Enabling_the_behavior_on_an_open_drop_down_aligns_it` | neuter `RunPass` |
| `Disabling_the_behavior_while_open_hands_the_offset_back` | neuter `RestoreBindingOffset` |

`Stepping_between_clamped_selections_still_re_aligns` earned its keep during this work: it caught an attempted fix that kept the layout listeners alive to confirm alignment, which would have made the drop-down override the user's own scrolling.

The Linux theme loads headlessly without an accent fallback and has no Liquid Glass/classic split, so none of the macOS test scaffolding was needed. Full fast suite: 48 passed.

## Verified on device

The headless host does not model popup placement constraints or compositor timing, so the remaining risk was checked by hand:

- Long lists (1000+ items) align correctly on first open and on reopen, on Linux under a Wayland session (XWayland, 100% scale).
- Screen-edge behaviour: no flipping; the popup slides and stays anchored.
- Display scaling: checked on macOS across Retina and non-Retina displays.
- Windows and macOS re-checked for regressions, since the behavior is shared.

The timing constants (60 ms stuck confirmation, 120 ms settle check) are the one judgement call: they were validated on a modestly-specced VM, and the failure mode if a machine were slower still is a drop-down row that is not perfectly aligned.

Baselines are not updated here; no visual-regression changes are expected, since the demo page already carries the ComboBoxes added in #652.

_Posted by Copilot CLI (agent), on behalf of @apman._

### CODE REVIEW

**Comment by apman** [comment]
Found it — this is a Wayland-specific timing bug, and it's a new one, not a recurrence of anything already fixed.

Wayland's popup placement (`xdg_popup`) is asynchronous: the compositor slides the popup to its final, on-screen position via a `configure` round-trip that can land *after* this behavior's own layout measurement. The correction loop was concluding "the popup is stuck against the screen edge, switch to scrolling the list" after just **one** repeated misalignment reading. On the first open of a far-down selection, that first repeat could be measuring the pre-slide position rather than a genuine clamp — so it switched to scrolling too early and settled a few rows off. On the second open, the window/surface state had already settled from the first pass, so even the very first measurement was already correct — which matches exactly what was reported: consistently wrong on open 1, consistently right from open 2 onward, with a variable-sized error (3 rows for selection 926, 5 rows for 905).

Fixed in `4bb3c4f6` by requiring three consecutive identical readings before concluding the popup is genuinely stuck, instead of one.

No test — the headless overlay popup host repositions synchronously, so there is no in-flight compositor state to reproduce in-process; two identical readings there really do mean stuck. This needs re-verifying on the Wayland VM.

_Posted by Copilot CLI (agent), on behalf of @apman._

**Comment by apman** [comment]
Pushed a real fix for the Wayland misalignment (the stuck-streak change only masked the symptom).

**Root cause:** on Wayland, `PointToScreen`/`PointToClient` are identity passthroughs — a client can't query its own window's absolute screen position, so `Position` is always `(0, 0)`. The alignment measurement was comparing "screen" points between the ComboBox's top level and the popup's top level, i.e. two unrelated coordinate systems. That's a wrong answer on every single pass, not a stale one that a retry would settle — which is exactly why raising the stuck-streak threshold only changed where the popup ended up wrong, not whether it did.

**Fix:** `Misalignment()` no longer touches screen coordinates at all. It translates both the ComboBox's and the selected row's centre into the popup's own root via `Visual.TranslatePoint`, then just compares the two points in that shared frame. That's exact by construction on every platform, and it also sidesteps a fixed few-pixel discrepancy an earlier attempt at this local-coordinates approach introduced by translating into `Popup.Child` instead of the popup's actual root (the template wraps content in a Border with a per-theme shadow margin).

Reverted `StuckStreakThreshold` back to 1 — the async Wayland "slide" that motivated raising it doesn't affect a local measurement like this.

All 48 fast tests pass, and I mutation-tested the new `Misalignment()` (neutered it, confirmed 8 tests fail) to make sure the assertions are actually meaningful.

Could you re-verify on the Wayland VM: first-open alignment for a selection far down a long list (e.g. reopening after picking ~926), and that the popup no longer briefly appears below the ComboBox / off-screen on first open?

_Posted by Copilot CLI (agent), on behalf of @apman._

**Comment by apman** [comment]
Found the real bug, and it's an important one to flag: the previous commit (60e93897) was built on a false premise.

I had earlier reasoned that Wayland's `Position`/`PointToScreen`/`PointToClient` are identity passthroughs, and rewrote `Misalignment()` to avoid them entirely using `Visual.TranslatePoint`. That reasoning turned out to rest on fabricated/hallucinated source files from an earlier debugging session (a local clone that had synthetic Wayland backend files spliced into it) — real Avalonia has **no separate Wayland backend at all**. On Linux, even under a Wayland session, the app runs through `Avalonia.X11` (via XWayland), and X11's `PointToScreen`/`PointToClient` are **not** identity — they track the real window position, updated via `ConfigureNotify` events.

The actual bug in 60e93897: `Visual.TranslatePoint` requires the two visuals to share a common ancestor reachable by walking `VisualParent` pointers. The popup's root and the ComboBox's own window are two separate native windows, linked only by a *logical* parent (used for focus routing), not a visual-tree one. So `TranslatePoint` silently returned `null` for the real popup window, and `Misalignment()`'s null guard made it return `0` — no correction, ever. That matches exactly what was reported: the error was never corrected even on reopening. It only appeared to work in the headless test suite because headless popups default to an overlay layer sharing one visual tree with the main window, where `TranslatePoint` trivially succeeds — a real gap in what those tests can catch.

Fix pushed in 7413a8e5:
- Reverted `Misalignment()` to `PointToScreen`/`PointToClient`, the same approach the macOS behavior already uses successfully.
- Restored `StuckStreakThreshold` to 3 (from 1), since a `PointToScreen`-based measurement is subject to the same asynchronous `ConfigureNotify` round-trip on X11/XWayland that motivated that threshold for macOS originally.

All 48 headless tests still pass. Could you re-verify on the Wayland VM again? Specifically: far-down selection reopening (should self-correct within a couple of frames, same as macOS), and the '1000 items' list edge case. Apologies for the wasted round-trip — should have verified the Wayland backend claim against the real Avalonia source from the start.

_Posted by Copilot CLI (agent), on behalf of @apman._

File: src/Devolutions.AvaloniaControls/Behaviors/ComboBoxPopupAlignmentBehavior.cs [hunk 1]
```diff
@@ -245,11 +306,50 @@ private void OnPopupOpened(object? sender, EventArgs e)
...
+            DispatcherTimer.RunOnce(
+                () =>
+                {
+                    if (scheduledFor != this.generation || !this.Popup.IsOpen) return;
+
+                    if (this.ResolveSelectedContainer() is not { } container) return;
+                    double misalignment = this.Misalignment(container);
+                    if (Math.Abs(misalignment) <= Tolerance) return;
+
+                    this.ReArm();
```
[the remark this answers is not in the document]
Comment (apman) [reply src/Devolutions.AvaloniaControls/Behaviors/ComboBoxPopupAlignmentBehavior.cs:378]: Good catch - fixed in de33f195, taking the first of your two suggestions.

The settle check now captures the scroller's offset when it is scheduled and abandons if it changed by the time it runs. A late compositor move repositions the popup without touching the list, so a changed scroll offset means the user wheeled it, and the check leaves the drop-down alone.

One thing worth recording about the regression test, since it caught me out. I first wrote the obvious version - scroll, wait out the delay, assert the row stayed put - and it passed **with the guard removed**. `DispatcherTimer` does not fire under the headless dispatcher at all (it runs on a virtual clock that sleeping does not advance), which I confirmed with a throwaway probe, so the callback under test never ran and the test was vacuous.

So the decision is now a separate `UserScrolledSince` method, covered by a theory that exercises it directly: unchanged, wheeled down, wheeled up, sub-tolerance jitter, and the two null cases. Neutering it to `=> false` fails two of the six. A second test guards the premise, asserting a scroll really does displace the row and nothing puts it back on its own.

_Posted by Copilot CLI (agent), on behalf of @apman._

File: src/Devolutions.AvaloniaControls/Behaviors/ComboBoxPopupAlignmentBehavior.cs [hunk 2]
```diff
@@ -213,8 +287,11 @@ private void ReArm()
 
             this.pass = 0;
             this.previousMisalignment = double.NaN;
+            this.stuckStreak = 0;
+            this.settleVerified = false;
             this.popupRoot = root;
             root.LayoutUpdated += this.OnLayoutUpdated;
+            if (root is WindowBase popupWindow) popupWindow.PositionChanged += this.OnPopupRootPositionChanged;
```
[the remark this answers is not in the document]
Comment (apman) [reply src/Devolutions.AvaloniaControls/Behaviors/ComboBoxPopupAlignmentBehavior.cs:300]: I checked this one rather than acting on it, and I don't think it holds: `PopupRoot` **is** a `WindowBase`, so `is WindowBase` already matches the native popup root.

Upstream (`src/Avalonia.Controls/Primitives/PopupRoot.cs`):

```csharp
public sealed class PopupRoot : WindowBase, IHostedVisualTreeRoot, IDisposable, IStyleHost, IPopupHost
```

I confirmed it against the actual assembly too, with a throwaway reflection probe: `typeof(WindowBase).IsAssignableFrom(typeof(PopupRoot))` is true, and `PositionChanged` has `DeclaringType == typeof(WindowBase)` - it is inherited, not shadowed, so subscribing via the base type reaches the same event.

There is also indirect evidence from the device testing: `PositionChanged` demonstrably *does* fire here. The bug this PR ends on was a stale reading being consumed by an extra pass, which is only reachable because position updates arrive asynchronously and drive the loop.

`PositionedPopupBehavior` matching on `PopupRoot` is not evidence against this - it needs the narrower type because it cares specifically about popup roots, whereas this behavior deliberately accepts any `WindowBase` so the same code path also covers the overlay/window hosts used in the headless tests. Narrowing it to `PopupRoot` would actually *lose* that coverage.

Leaving the code as is. Happy to revisit if you have a case where `TopLevel.GetTopLevel` returns a popup host that is not a `WindowBase`.

_Posted by Copilot CLI (agent), on behalf of @apman._

File: src/Devolutions.AvaloniaControls/Behaviors/ComboBoxPopupAlignmentBehavior.cs [hunk 3]
```diff
@@ -273,11 +400,42 @@ private void Unsubscribe()
...
+        /// millisecond, which is what let a stuck streak build up entirely from stale, pre-move
+        /// readings (see <see cref="StuckStreakThreshold"/>).
+        /// </summary>
+        private void ScheduleStuckConfirmation()
+        {
+            int scheduledFor = this.generation;
+            DispatcherTimer.RunOnce(
+                () =>
+                {
+                    if (scheduledFor != this.generation) return;
```
[the remark this answers is not in the document]
Comment (apman) [reply src/Devolutions.AvaloniaControls/Behaviors/ComboBoxPopupAlignmentBehavior.cs:448]: Agreed, and fixed in 6efc01d8.

`SchedulePass` already coalesces its own timer via `passScheduled`; `ScheduleStuckConfirmation` had no equivalent, so `LayoutUpdated` and `PositionChanged` re-entering `RunPass` within the 60 ms window could each queue one. `ScheduleStuckConfirmation` now guards on a `stuckConfirmationScheduled` flag, cleared in the callback and in `Unsubscribe` - which is the single choke point every lifecycle reset goes through (`ReArm`, `OnPopupOpened`, `OnPopupClosed`, `Dispose`), so it cannot leak a stale `true` into the next opening.

No test for this one, deliberately: `DispatcherTimer` does not fire under the headless dispatcher (established on the previous thread), so anything asserting on multiple queued timers would be vacuous. The guard is a direct structural mirror of the existing `passScheduled` one.

_Posted by Copilot CLI (agent), on behalf of @apman._

File: src/Devolutions.AvaloniaControls/Devolutions.AvaloniaControls.csproj [hunk 4]
```diff
@@ -25,6 +25,11 @@
         </PackageReleaseNotes>
     </PropertyGroup>
 
+    <!-- Debug-only: lets the tests reach internals, without shipping the attribute in the package. -->
+    <ItemGroup Condition="'$(Configuration)' == 'Debug'">
+        <InternalsVisibleTo Include="Devolutions.AvaloniaControls.VisualTests" />
+    </ItemGroup>
```
[the remark this answers is not in the document]
Comment (apman) [reply src/Devolutions.AvaloniaControls/Devolutions.AvaloniaControls.csproj:31]: I checked this and the premise does not hold: the test project cannot be built in Release at all, on this branch or on `master`.

```
CSC : error CS1704: An assembly with the same simple name 'Devolutions.AvaloniaControls'
has already been imported. [.../samples/SampleApp/SampleApp.csproj]
```

That is the same mixed package-plus-project reference situation you describe, but it already breaks the *build* well before any friend-assembly check, and it reproduces unchanged on `master` with none of this PR's commits. So a Release build of the tests is not a configuration that exists today, and the `Debug`-only friend declaration does not remove one that worked.

The `Debug` condition also matches how this csproj already handles the same distinction - the `ProjectReference`/`PackageReference` split for `Devolutions.AvaloniaControls` is conditioned on exactly the same property. Tests build and run against the project reference in `Debug`, which is where the internal is reachable, and the attribute stays out of the shipped package.

Happy to revisit if Release builds of the test project are meant to work - but that would be a pre-existing issue to fix on its own, not something this PR introduced.

_Posted by Copilot CLI (agent), on behalf of @apman._

File: src/Devolutions.AvaloniaTheme.Linux/Controls/ComboBox.axaml [hunk 5]
```diff
@@ -128,6 +129,9 @@
     <Setter Property="FocusAdorner" Value="{x:Null}" />
     <Setter Property="ClipToBounds" Value="False" />
     <Setter Property="ComboBoxValidationBehavior.Enable" Value="True" />
+    <!-- Corrects the VerticalOffset estimate below against the popup's real layout, so the selected
+         row still lands on the ComboBox once the option list is long enough to scroll. -->
+    <Setter Property="ComboBoxPopupAlignmentBehavior.Enable" Value="True" />
```
[the remark this answers is not in the document]
Comment (apman) [reply src/Devolutions.AvaloniaTheme.Linux/Controls/ComboBox.axaml:134]: Correct, and this is already the plan for release - thanks for flagging it independently.

One correction to the detail: `Devolutions.AvaloniaControls` **2026.9.3** is already published and does contain `ComboBoxPopupAlignmentBehavior` (verified in the resolved package). Because the range is open-ended, a normal restore picks that up, which is why non-Debug builds of both themes succeed today.

The exposure is the declared *minimum* rather than what restore happens to choose: `[2026.8.25,)` still permits 2026.8.25, where the attached property does not exist. Worth noting this is not new to this PR - the macOS theme has had the same setter, and the same declared minimum, since #652.

So the sequencing you describe is exactly right, and the maintainer has already scheduled it: publish Controls first, then raise `MinimumAvaloniaControlsVersion` in `Common.props` so both themes can be republished together. Keeping it as a release step rather than putting a speculative version bump in this PR, since the value depends on which Controls version actually ships.

_Posted by Copilot CLI (agent), on behalf of @apman._
